# 예외처리

## 목차
[1. 예외와 예외 클래스](#1-예외와-예외-클래스) </br>
[2. 예외 처리 코드](#2-예외-처리-코드) </br>
[3. 자동 리소스 닫기](#3-자동-리소스-닫기) </br>
[4. 사용자 저의 예외](#4-사용자-정의-예외)
___

## 1. 예외와 예외 클래스
컴퓨터 하드웨어와 오동작 또는 고장으로 인해 응용프로그램 실행 오류가 발생하는 것을 자바에서는 Error라고 한다. 자바에서는 에러 이외에 Exception라고 부르는 오류가 있다. 예외란 사용자의 잘못된 조작 또는 개발자의 잘못된 코딩으로 인해 발생하는 프로그램 오류를 말한다.

예외는 두 가지 종류가 있다.

1. 일반예외
- 일반 예외는 컴파일러 체크 예외라고도 하는데, 자바 소스를 컴파일하는 과정에서 예외 처리 코드가 필요한지 검사하기 떄문이다. 만약 예외 처리 코드가 없다면 컴파일 오류가 발생한다.

2. 실행 예외
- 실행 예외는 컴파일하는 과정에서 예외 처리 코드를 검사하지 않는 예외를 말한다. 컴파일 시 예외 처리를 확인하는 차이일 뿐, 두 가지 예외는 모두 예외처리가 필요하다.

![예외처리종류](./ExceptionClassHierarchy.png)
> 출처 : https://velog.io/@codepark_kr/%EC%9E%90%EB%B0%94-%EC%9D%B4%EB%A1%A0-%EC%98%88%EC%99%B8-%EC%B2%98%EB%A6%AC

일반 예외와 실행 예외 클래스를 구별하는 방법은 일반 예외는 Exception을 상ㅅ혹받지만 Runtime Exception은 상속받지 않는 클래스들이고, 실행 예외는 다음과 같이 RuntimeException을 상속받은 클래스들이다. __RuntimeException 역시 Exception을 상속받지만, JVM은 RuntimeException을 상속했는지 여부를 보고 실행 예외를 판단한다.__

## 1.1 실행 예외
실행 예외는 자바 컴파일러가 체크를 하지 않기 때문에 오로지 개발자의 경험에 의해서 예외 처리 코드를 삽입해야 한다. 만약 개발자가 실행 예외에 대해 예외 처리 코드를 넣지 않았을 경우, 해당 예외가 발생하면 프로그램은 곧바로 종료된다.
___

## 2. 예외 처리 코드
프로그램에서 예외가 발생했을 경우 프로그램의 갑작스러운 종료를 막고, 정상 실행을 유지할 수 있도록 처리하는 코드를 예외 처리 코드라고 한다. 실행 예외는 컴파일러가 체크해주지 않기 때문에 예외 처리 코드를 개발자의 경험을 바탕으로 작성해야 한다. 예외 처리 코드는 try-catch-finally 블록을 이용한다.

```java
try {
  // 예외 발생시 catch로 이동.
} catch (예외클래스 e) {
  // 예외 처리 후 finally로 이동.
} finally {
  // 항상 실행
}
```
___

## 3. 자동 리소스 닫기
자바 7에서 새로 추가된 try-with-resources를 사용하면 예외 발생 여부와 상관없이 사용했던 리소스 객체의 close() 메소드를 호출해서 안전하게 리소스를 닫아준다. 

```java
/**
* 수동으로 리소스를 해제할 경우
*/
FileInputStream fis = null;
try {
    fis = new FileInputStream("파일경로");
    ...
} catch(IOException e) {
    ...
} finally {
    if (fis != null) {
        try {
            fis.close();
        } catch (IOException e) {
            ...
        }
    }
}

/**
* try-with-resources로 해제할 경우
*/
try (FileInputStream fis = new FileInputStream("파일경로");
     FileOutputStream fos = new FileOutputStream("파일경로")) {
    ...
} catch(IOException e) {
    ...
}
```

try-with-resources를 사용하기 위해서는 조건이 있는데, 리소스 객체는 __java.lang.AutoCloseable 인터페이스__ 를 구현하고 있어야 한다. AutoCloseable에는 close() 메소드가 정의되어 있는데 try-with-resource는 바로 이 close() 메소드를 자동 호출한다.

## 3.1 예외 떠넘기기.
메소드 내부에서 예외가 발생할 수 있는 코드를 작성할 때 try-catch 블록으로 예외를 처리하는 것이 기본이지만, 경우에 따라서는 메소드를 호출한 곳으로 예외를 떠넘길 수도 있다. 이때 사용하는 키워드가 __throws__ 이다. throws 키워드는 메소드 선언부 끝에 작성되어 메소드에서 처리하지 않은 예외를 호출한 곳으로 떠넘기는 역할을 한다.

```java
public void method1() {
    try {
        method2();
    } catch (Exception e) {
        System.out.println("예외 발생.");
    }
}

public void method2() throws ClassNotFoundException {
    Class clazz = Class.forName("java.lang.String2");
}
```
___

## 4. 사용자 정의 예외
애플리케이션 예외는 개발자가 직접 정의해서 만들 수 있는데 이를 __사용자 정의 예외__ 라고 한다.

사용자 정의 예외 클래스는 컴파일러가 체크하는 일반 예외로 선언할 수도 있고, 컴파일러가 체크하지 않는 실행 예외로 선언할 수도 있다. 일반 예외로 선언할 경우 Exception을 상속하면 되고, 실행 예외로 선언할 경우에는 RuntimeException을 상속하면 된다.

```java
public Class ApiException extends [ Exception | RuntimeException ] {
    public ApiException() {}
    public ApiException(String message) { super(message); }
}
```

사용자 정의 예외 클래스 이름은 Exception으로 끝나는 것이 좋다. 사용자 정의 예외 클래스도 필드, 생성자, 메소드 선언들을 포함할 수 있지만 대부분 생성자 선언만을 포함한다. 생성자는 두 개를 선언하는 것이 일반적인데, 하나는 매개 변수가 없는 기본 생성자이고, 다른 하나는 예외 발생 원인 메세지를 전달하기 위해 String 타입의 매개 변수를 갖는 생성자이다.

> 참조 : 이것이 자바다(신용권의 Java 프로그래밍 정복)
